File: motor/procesar.py
"""Orquestador del pipeline completo (D6):

adaptador → limpieza → extracción de notas → imágenes/separadores → split → render."""

from pathlib import Path

from motor.adaptadores.calibre import documentos_calibre
from motor.adaptadores.docx import convertir_docx
from motor.adaptadores.markdown import _restaurar_imagenes, documentos_markdown
from motor.division import dividir_en_capitulos
from motor.imagenes import procesar_imagenes, procesar_separadores
from motor.limpieza import limpiar_texto_html
from motor.modelo import Chapter, Contadores, Nota, Resultado
from motor.notas import asignar_capitulos, extraer_notas


def _procesar_documento(
    html: str,
    notas: list[Nota],
    imagenes_placeholder: list[str] | None = None,
) -> tuple[str, int, int]:
    """Pipeline común por documento: notas → limpieza → imágenes → separadores.
    Las notas se extraen ANTES de la limpieza porque ésta elimina los
    atributos id=/class= de las anclas de pandoc (<section id="footnotes">)."""
    html, notas_doc = extraer_notas(html)
    notas.extend(notas_doc)
    html = limpiar_texto_html(html)

    n_imagenes = 0
    if imagenes_placeholder:
        html = _restaurar_imagenes(html, imagenes_placeholder)
        n_imagenes += len(imagenes_placeholder)

    html, n_img = procesar_imagenes(html)
    n_imagenes += n_img

    html, n_sep = procesar_separadores(html)
    return html, n_imagenes, n_sep


def procesar(
    modo: str,
    ruta_entrada: Path,
    ruta_template: Path,
    titulos: list[str] | None = None,
    start_num: int = 1,
) -> Resultado:
    """Procesa el manuscrito completo y devuelve capítulos, notas y conteos."""
    if modo not in ('word', 'calibre', 'markdown'):
        raise ValueError(f'Modo desconocido: {modo}')

    template = ruta_template.read_text(encoding='utf-8')
    if '{{CONTENIDO}}' not in template:
        raise ValueError('El template no contiene el placeholder {{CONTENIDO}}')

    notas: list[Nota] = []
    capitulos: list[Chapter] = []
    avisos: list[str] = []
    n_imagenes_total = 0
    n_separadores_total = 0

    if modo == 'word':
        html = convertir_docx(ruta_entrada)
        html, n_img, n_sep = _procesar_documento(html, notas)
        n_imagenes_total += n_img
        n_separadores_total += n_sep
        capitulos = dividir_en_capitulos(html, titulos, start_num)

    else:
        if modo == 'calibre':
            documentos, avisos_doc = documentos_calibre(ruta_entrada)
        else:
            documentos, avisos_doc = documentos_markdown(ruta_entrada)
        avisos.extend(avisos_doc)
        for indice, (html_doc, titulo_auto, imagenes_placeholder) in enumerate(documentos):
            html, n_img, n_sep = _procesar_documento(html_doc, notas, imagenes_placeholder)
            n_imagenes_total += n_img
            n_separadores_total += n_sep

            num = start_num + indice
            titulo = (
                titulos[indice]
                if titulos and indice < len(titulos)
                else (titulo_auto or f'Capítulo {num}')
            )
            capitulos.append(Chapter(titulo=titulo, html_cuerpo=html.strip()))

    asignar_capitulos(notas, capitulos, start_num)

    contadores = Contadores(
        capitulos=len(capitulos),
        notas=len(notas),
        imagenes=n_imagenes_total,
        separadores=n_separadores_total,
    )
    return Resultado(capitulos=capitulos, notas=notas, contadores=contadores, avisos=avisos)

File: app_flutter/lib/motor/procesar.dart
import 'dart:io';
import 'modelo.dart';
import 'limpieza.dart';
import 'notas.dart';
import 'imagenes.dart';
import 'division.dart';
import 'adaptadores.dart';
import 'plantillas.dart';

typedef _ProcesarDocumentoResult = ({String html, int nImagenes, int nSeparadores});

_ProcesarDocumentoResult _procesarDocumento(String html, List<Nota> notas, {List<String>? imagenesPlaceholder}) {
  final extraerResult = extraerNotas(html);
  html = extraerResult.html;
  notas.addAll(extraerResult.notas);
  
  html = limpiarTextoHtml(html);

  int nImagenes = 0;
  if (imagenesPlaceholder != null && imagenesPlaceholder.isNotEmpty) {
    html = restaurarImagenesMarkdown(html, imagenesPlaceholder);
    nImagenes += imagenesPlaceholder.length;
  }

  final procesarImgResult = procesarImagenes(html);
  html = procesarImgResult.html;
  nImagenes += procesarImgResult.count;

  final procesarSepResult = procesarSeparadores(html);
  html = procesarSepResult.html;
  int nSeparadores = procesarSepResult.count;

  return (html: html, nImagenes: nImagenes, nSeparadores: nSeparadores);
}

Future<Resultado> procesar({
  required String modo,
  required String rutaEntrada,
  required String rutaTemplate,
  List<String>? titulos,
  int startNum = 1,
  String? rutaPlantillas,
}) async {
  if (!['word', 'calibre', 'markdown', 'pdf'].contains(modo)) {
    throw Exception('Modo desconocido: $modo');
  }

  String template = File(rutaTemplate).readAsStringSync();
  if (!template.contains('{{CONTENIDO}}')) {
    throw Exception('El template no contiene el placeholder {{CONTENIDO}}');
  }

  List<Nota> notas = [];
  List<Chapter> capitulos = [];
  List<String> avisos = [];
  int nImagenesTotal = 0;
  int nSeparadoresTotal = 0;

  if (modo == 'word' || modo == 'pdf') {
    String htmlRaw;
    if (modo == 'word') {
      htmlRaw = await convertirDocx(rutaEntrada);
    } else {
      htmlRaw = await convertirPdf(rutaEntrada);
    }
    
    final procResult = _procesarDocumento(htmlRaw, notas);
    nImagenesTotal += procResult.nImagenes;
    nSeparadoresTotal += procResult.nSeparadores;
    
    capitulos = dividirEnCapitulos(procResult.html, titulos: titulos, startNum: startNum);
    for (var cap in capitulos) {
      cap.htmlRaw = htmlRaw;
    }
  } else {
    DocumentosResult resultDocs;
    if (modo == 'calibre') {
      resultDocs = documentosCalibre(rutaEntrada);
    } else {
      resultDocs = await documentosMarkdown(rutaEntrada);
    }
    avisos.addAll(resultDocs.avisos);
    
    for (int indice = 0; indice < resultDocs.documentos.length; indice++) {
      final doc = resultDocs.documentos[indice];
      final procResult = _procesarDocumento(doc.html, notas, imagenesPlaceholder: doc.imagenesPlaceholder);
      nImagenesTotal += procResult.nImagenes;
      nSeparadoresTotal += procResult.nSeparadores;

      int num = startNum + indice;
      String titulo = (titulos != null && indice < titulos.length) ? titulos[indice] : (doc.titulo ?? 'Capítulo $num');
      capitulos.add(Chapter(titulo: titulo, htmlCuerpo: procResult.html.trim(), htmlRaw: doc.html));
    }
  }

  clasificarYRenumerarCapitulos(capitulos, startNum: startNum, rutaPlantillas: rutaPlantillas);

  asignarCapitulos(notas, capitulos, startNum: startNum);

  return Resultado(
    capitulos: capitulos,
    notas: notas,
    contadores: Contadores(
      capitulos: capitulos.length,
      notas: notas.length,
      imagenes: nImagenesTotal,
      separadores: nSeparadoresTotal,
    ),
    avisos: avisos,
  );
}
